runtime/lua/vim/snippet.lua

```diff
@@ -0,0 +1,212 @@
+    if node.children then
+      for _, child in ipairs(node.children) do
+        for _, mark in ipairs(traverse(child, end_pos)) do
+          table.insert(marks, mark)
+        end
+      end
+    else
+      local texts = vim.split(tostring(node), '\n')
+      end_pos[1] = start_pos[1] + #texts - 1
+      end_pos[2] = #texts == 1 and (start_pos[2] + #texts[1]) or #texts[#texts]
+    end
+
+    if vim.tbl_contains({
+      snippet.NodeType.TABSTOP,
+      snippet.NodeType.PLACEHOLDER,
+      snippet.NodeType.CHOICE,
+    }, node.type) then
+      table.insert(marks, {
+        mark_id = vim.api.nvim_buf_set_extmark(0, self.ns, start_pos[1], start_pos[2], {
+          end_line = end_pos[1],
+          end_col = end_pos[2],
+          hl_group = 'Visual',
+        }),
+        node = node,
+      })
+    end
+
+    start_pos[1] = end_pos[1]
+    start_pos[2] = end_pos[2]
+
+    return marks
+  end
+  self.marks = traverse(snippet_node, cursor)
+end
+
+---Jump to the suitable placeholder.
+---@param direction nvim.SnippetJumpDirection
+function SnippetSession:jump(direction)
+  local is_correct_direction = function(cursor, current_range)
+    if direction == JumpDirection.Next then
+      return current_range.s[1] < cursor[1] or (current_range.s[1] == cursor[1] and cursor[2] < current_range.s[2])
+    end
+    return current_range.s[1] > cursor[1] or (current_range.s[1] == cursor[1] and cursor[2] > current_range.s[2])
+  end
+  local compare = function(cursor, target_range, current_range)
+    local is_nearest = not target_range
+    if direction == JumpDirection.Next then
+      is_nearest = is_nearest or (math.abs(target_range.s[1] - cursor[1]) > math.abs(current_range.s[1] - cursor[1]))
+      is_nearest = is_nearest or (target_range.s[1] == current_range.s[1] and math.abs(target_range.s[2] - cursor[2]) > math.abs(current_range.s[2] - cursor[2]))
+    else
+      is_nearest = is_nearest or (math.abs(target_range.s[1] - cursor[1]) < math.abs(current_range.s[1] - cursor[1]))
+      is_nearest = is_nearest or (target_range.s[1] == current_range.s[1] and math.abs(target_range.s[2] - cursor[2]) < math.abs(current_range.s[2] - cursor[2]))
+    end
+    return is_nearest
+  end
+
+  local cursor = vim.api.nvim_win_get_cursor(0)
+  cursor[1] = cursor[1] - 1
+
+  local target_range
+  for _, m in ipairs(state.session.marks or {}) do
+    local mark = vim.api.nvim_buf_get_extmark_by_id(0, state.session.ns, m.mark_id, { details = true })
+    local s = { mark[1], mark[2] }
+    local e = { mark[3].end_row, mark[3].end_col }
+    if s[1] > e[1] or (s[1] == e[1] and s[2] > e[2]) then
+      local t = s
+      s = e
+      e = t
+    end
+    local current_range = { s = s, e = e }
+    if is_correct_direction(cursor, current_range) then
+      if compare(cursor, target_range, current_range) then
+        target_range = current_range
+      end
+    end
+  end
+  if target_range then
+    select_text(target_range)
+  end
+end
+
+local M = {}
```

Comment:
This `M` will be a `vim.snippet`.

The plugin authors can override `vim.snippet.expand` and `vim.snippet.jump`.

Probably, we should expose `vim.snippet.jumpable(direction)`.